source/agora/consensus/protocol/Nominator.d

```diff
@@ -577,7 +578,14 @@ extern(D):
             return;  // slot was already externalized or envelope is too new
         }
 
-        const PublicKey public_key = PublicKey(envelope.statement.nodeID[]);
+        const Hash utxo = Hash(envelope.statement.nodeID[][0 .. Hash.sizeof]);
```

Comment:
The (double) slicing is not necessary here.

For `NodeID` we need to do:
`auto node_id = NodeID(self_enroll[][0 .. NodeID.sizeof]);`
Because:
- `self_enroll` is a `Hash` (BitBlob);
- `NodeID` expects a `ubyte[64]`

We can get a `ubyte[]` out of a `BitBlob`, by slicing.
Then, you can get a static array out of a dynamic array if you slice it with compile-time-known values (such as `Hash.sizeof`). So `self_enroll[]` (`ubyte[]`) => `self_enroll[][0 .. 64]`, the type of this expression is *still* `ubyte[]` **but** the compiler knows it can be converted to a static array. It's a form of value-range propagation/analysis. You have the same with literals: `auto x = 42;`, the type of `x` is `int`. But if you do `ubyte x = 42;`, it works, despite `int i; ubyte x = i;` not working.

However, the constructor of `BitBlob` [accepts a slice](https://github.com/Geod24/bitblob/blob/cbf4b05c834fd8b8fb06457d287613dbe7f74b74/source/geod24/bitblob.d#L198), so there is no such requirement as with `NodeID`.

The easiest way to go about it would probably be to allow `BitBlob!X` => `ubyte[X]` conversion, of course, which would remove our need to have this weird slicing statements.

Reply:
Ty for the great explanation. :)

Reply:
One more question. So is `const Hash utxo = envelope.statement.nodeID;` enough?